File: packages/browseros-agent/packages/browser-mcp/src/tools/windows.ts
import type { WindowInfo } from '@browseros/browser-core/core/windows'
import { z } from 'zod/v4'
import { defineTool, errorResult, textResult } from './framework'

const ACTIONS = ['list', 'create', 'close', 'activate'] as const

export const windows = defineTool({
  name: 'windows',
  description:
    'Manage browser windows: list, create, close, or activate a window.',
  input: z
    .object({
      action: z.enum(ACTIONS).default('list'),
      windowId: z
        .number()
        .int()
        .optional()
        .describe('Window id for close and activate.'),
    })
    .strict(),
  annotations: {
    title: 'Manage windows',
    destructiveHint: true,
    openWorldHint: true,
  },
  handler: async (args, ctx) => {
    switch (args.action) {
      case 'list': {
        const all = await ctx.session.windows.list()
        return textResult(formatWindowList(all), {
          action: 'list',
          windows: all,
          count: all.length,
        })
      }
      case 'create': {
        const window = await ctx.session.windows.create()
        return textResult(`created window ${window.windowId}`, {
          action: 'create',
          window,
        })
      }
      case 'close': {
        if (args.windowId === undefined) {
          return errorResult('windows close: windowId is required.')
        }
        await ctx.session.windows.close(args.windowId)
        return textResult(`closed window ${args.windowId}`, {
          action: 'close',
          windowId: args.windowId,
        })
      }
      case 'activate': {
        if (args.windowId === undefined) {
          return errorResult('windows activate: windowId is required.')
        }
        await ctx.session.windows.activate(args.windowId)
        return textResult(`activated window ${args.windowId}`, {
          action: 'activate',
          windowId: args.windowId,
        })
      }
      default:
        return errorResult('windows: unsupported action.')
    }
  },
})

function formatWindowList(windows: WindowInfo[]): string {
  if (windows.length === 0) return 'No windows found.'

  const lines = [`Found ${windows.length} windows:`, '']
  for (const window of windows) {
    const suffix = window.isActive ? ' [ACTIVE]' : ''
    lines.push(
      `Window ${window.windowId} (${window.windowType}, ${window.tabCount} tabs)${suffix}`,
    )
  }
  return lines.join('\n')
}

File: packages/browseros-agent/crates/browseros-mcp/src/tools/windows.rs
use crate::framework::{
    ToolCtx, ToolExecResult, ToolResult, error_result, parse_args, text_result,
};
use browseros_core::WindowId;
use futures_util::future::BoxFuture;
use schemars::JsonSchema;
use serde::Deserialize;
use serde_json::{Value, json};

const DESCRIPTION: &str = "Manage browser windows: list, create, or close windows.";

#[derive(Debug, Clone, Default, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
enum WindowsAction {
    #[default]
    List,
    Create,
    Close,
    // Intentionally omit Activate so MCP callers cannot steal window focus.
}

#[derive(Debug, Clone, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
struct WindowsArgs {
    #[serde(default)]
    action: WindowsAction,
    /// Window id for close.
    #[serde(rename = "windowId")]
    window_id: Option<i64>,
}

pub fn definition() -> crate::framework::ToolDef {
    super::def::<WindowsArgs>(
        "windows",
        DESCRIPTION,
        Some(super::open_world_annotations()),
        handler,
    )
}

fn handler<'a>(
    raw: Value,
    ctx: &'a ToolCtx,
    _response: &'a mut crate::response::ToolResponse,
) -> BoxFuture<'a, ToolExecResult<Option<ToolResult>>> {
    Box::pin(async move {
        let args: WindowsArgs = parse_args(raw)?;
        let result = match args.action {
            WindowsAction::List => {
                let windows = ctx.session.windows.list().await?;
                text_result(
                    format_window_list(&windows),
                    Some(json!({ "action": "list", "windows": windows, "count": windows.len() })),
                )
            }
            WindowsAction::Create => {
                let window = ctx.session.windows.create().await?;
                text_result(
                    format!("created window {}", window.window_id),
                    Some(json!({ "action": "create", "window": window })),
                )
            }
            WindowsAction::Close => {
                let Some(window_id) = args.window_id else {
                    return Ok(Some(error_result("windows close: windowId is required.")));
                };
                ctx.session.windows.close(WindowId(window_id)).await?;
                text_result(
                    format!("closed window {window_id}"),
                    Some(json!({ "action": "close", "windowId": window_id })),
                )
            }
        };
        Ok(Some(result))
    })
}

fn format_window_list(windows: &[browseros_core::windows::WindowInfo]) -> String {
    if windows.is_empty() {
        return "No windows found.".to_string();
    }
    let mut lines = vec![format!("Found {} windows:", windows.len()), String::new()];
    for window in windows {
        let mut markers = Vec::new();
        if !window.is_visible {
            markers.push("NOT VISIBLE");
        }
        if window.is_active {
            markers.push("ACTIVE");
        }
        let suffix = if markers.is_empty() {
            String::new()
        } else {
            format!(" [{}]", markers.join(", "))
        };
        lines.push(format!(
            "Window {} ({}, {} tabs){suffix}",
            window.window_id, window.window_type, window.tab_count
        ));
    }
    lines.join("\n")
}
